src/anonymeter/evaluators/singling_out_evaluator.py

```diff
@@ -332,14 +331,25 @@ def multivariate_singling_out_queries(df: pd.DataFrame, n_queries: int, n_cols:
     so_queries = UniqueSinglingOutQueries()
     medians = df.median(numeric_only=True)
 
+    n_attempts = 0
+
     while len(so_queries) < n_queries:
+        if max_attempts is not None and n_attempts >= max_attempts:
+            logger.warning(
+                "Reached maximum number of attempts when generating singling out queries. "
+                f"Returning {len(so_queries.queries)} instead of the requested {n_queries}."
+            )
+            return so_queries.queries
+
         record = df.iloc[rng.integers(df.shape[0])]
         columns = rng.choice(df.columns, size=n_cols, replace=False).tolist()
 
         query = _query_from_record(record=record, dtypes=df.dtypes, columns=columns, medians=medians)
 
         so_queries.check_and_append(query=query, df=df)
```

Comment:
I might be confused a little here: my understanding that if the generated query is "good" and will be appended - the counter needs to be reset. The pseudocode:
```
query = _query_from_record(...)
if check(query):
    n_attempts = 0
    append(query)
else:
    n_attempts += 1
```
And if this is the case, then this counter could be moved inside of the `UniqueSinglingOutQueries` class.

Otherwise it might be a bit confusing that the attempts are increasing even if the generated queries are "good".

Reply:
I see your point.. For me the attempts are the total attempts, not the "per-query one". So the attacker has, say `max_attempts` attempts to generate `n_attack` queries. Even a successful generation will count. 

Your solution is also valid, and I think that the results would be identical (no infinite loop). Unless you fell strongly about this, I'll stick with the global attempts parameter. Let me know what you think.

Reply:
I'm fine with the global parameter, I'd just explain in the docstring more clearly what it means. I'd assume others might get confused just like I did.